File: tests/python/q1_hidden.py
from solutions.python.question1 import ParcelLockerService


def exact_limits() -> None:
    service = ParcelLockerService()
    assert service.add_locker("L", 5) is True
    assert service.store_parcel(1, "x", "L", 5) is True
    assert service.get_load(2, "L") == 5
    assert service.collect_parcel(3, "x") is True
    assert service.collect_parcel(4, "x") is False
    assert service.store_parcel(5, "", "L", 1) is False
    assert service.store_parcel(6, "y", "L", -1) is False


def shared_capacity() -> None:
    service = ParcelLockerService()
    service.add_locker("L", 7)
    assert service.store_parcel(1, "a", "L", 3) is True
    assert service.store_parcel(2, "b", "L", 4) is True
    assert service.store_parcel(3, "c", "L", 1) is False
    assert service.collect_parcel(4, "a") is True
    assert service.store_parcel(5, "c", "L", 3) is True
    assert service.get_load(6, "L") == 7


def ranking_tie_and_limit() -> None:
    service = ParcelLockerService()
    service.add_locker("z", 5)
    service.add_locker("a", 5)
    service.add_locker("m", 5)
    assert service.top_lockers(1, 2) == ["a(0)", "m(0)"]
    assert service.top_lockers(2, 0) == []


def invalid_moves_do_not_score() -> None:
    service = ParcelLockerService()
    service.add_locker("A", 5)
    service.store_parcel(1, "p", "A", 2)
    assert service.move_parcel(2, "p", "A") is False
    assert service.move_parcel(3, "p", "X") is False
    assert service.move_parcel(4, "missing", "A") is False
    assert service.top_lockers(5, 1) == ["A(2)"]


def scheduled_numeric_order() -> None:
    service = ParcelLockerService()
    service.add_locker("A", 50)
    service.add_locker("B", 50)
    for number in range(1, 12):
        service.store_parcel(number, f"p{number}", "A", 1)
        assert (
            service.schedule_move(20 + number, f"p{number}", "B", 100)
            == f"move-{number}"
        )
    assert service.process_scheduled(100) == [f"move-{number}" for number in range(1, 12)]


def scheduled_capacity_failure() -> None:
    service = ParcelLockerService()
    service.add_locker("A", 10)
    service.add_locker("B", 3)
    service.store_parcel(1, "p", "A", 4)
    assert service.schedule_move(2, "p", "B", 5) == "move-1"
    assert service.process_scheduled(5) == []
    assert service.process_scheduled(6) == []
    assert service.collect_parcel(7, "p") is True
    assert service.cancel_scheduled_move(8, "move-1") is False


def merge_capacity_failure() -> None:
    service = ParcelLockerService()
    service.add_locker("A", 5)
    service.add_locker("B", 5)
    service.store_parcel(1, "a", "A", 4)
    service.store_parcel(2, "b", "B", 3)
    assert service.merge_lockers(3, "A", "B") is False
    assert service.get_load(4, "A") == 4
    assert service.get_load(5, "B") == 3
    assert service.top_lockers(6, 2) == ["A(4)", "B(3)"]
    assert service.get_parcel_history(7, "a") == ["stored:A@1"]


def merge_retargets_pending() -> None:
    service = ParcelLockerService()
    service.add_locker("A", 10)
    service.add_locker("B", 10)
    service.add_locker("C", 10)
    service.store_parcel(1, "p", "C", 2)
    assert service.schedule_move(2, "p", "A", 10) == "move-1"
    assert service.merge_lockers(3, "A", "B") is True
    assert service.process_scheduled(10) == ["move-1"]
    assert service.get_load(11, "B") == 2
    assert service.get_parcel_history(12, "p") == [
        "stored:C@1",
        "moved:C->B@10",
    ]


CASES = [
    (1, "limites exatos e coleta repetida", exact_limits),
    (1, "capacidade é compartilhada", shared_capacity),
    (2, "desempata ranking e respeita n", ranking_tie_and_limit),
    (2, "movimentos inválidos não alteram atividade", invalid_moves_do_not_score),
    (3, "processa empate por ordem numérica de criação", scheduled_numeric_order),
    (3, "falha agendada por capacidade é definitiva", scheduled_capacity_failure),
    (4, "fusão sem capacidade falha atomicamente", merge_capacity_failure),
    (4, "fusão redireciona agendamento pendente", merge_retargets_pending),
]

File: tests/typescript/q1.hidden.ts
import { ParcelLockerService } from "../../solutions/typescript/question1.js";
import { equal, type TestCase } from "../../grader/typescript/harness.js";

export const hiddenCases: TestCase[] = [
  {
    level: 1,
    name: "limites exatos e coleta repetida",
    run: () => {
      const service = new ParcelLockerService();
      equal(service.addLocker("L", 5), true);
      equal(service.storeParcel(1, "x", "L", 5), true);
      equal(service.getLoad(2, "L"), 5);
      equal(service.collectParcel(3, "x"), true);
      equal(service.collectParcel(4, "x"), false);
      equal(service.storeParcel(5, "", "L", 1), false);
      equal(service.storeParcel(6, "y", "L", -1), false);
    }
  },
  {
    level: 1,
    name: "capacidade é compartilhada entre várias encomendas",
    run: () => {
      const service = new ParcelLockerService();
      service.addLocker("L", 7);
      equal(service.storeParcel(1, "a", "L", 3), true);
      equal(service.storeParcel(2, "b", "L", 4), true);
      equal(service.storeParcel(3, "c", "L", 1), false);
      equal(service.collectParcel(4, "a"), true);
      equal(service.storeParcel(5, "c", "L", 3), true);
      equal(service.getLoad(6, "L"), 7);
    }
  },
  {
    level: 2,
    name: "desempata lexicograficamente e respeita n",
    run: () => {
      const service = new ParcelLockerService();
      service.addLocker("z", 5);
      service.addLocker("a", 5);
      service.addLocker("m", 5);
      equal(service.topLockers(1, 2), ["a(0)", "m(0)"]);
      equal(service.topLockers(2, 0), []);
    }
  },
  {
    level: 2,
    name: "recusa destino atual e inexistente sem alterar métricas",
    run: () => {
      const service = new ParcelLockerService();
      service.addLocker("A", 5);
      service.storeParcel(1, "p", "A", 2);
      equal(service.moveParcel(2, "p", "A"), false);
      equal(service.moveParcel(3, "p", "X"), false);
      equal(service.moveParcel(4, "missing", "A"), false);
      equal(service.topLockers(5, 1), ["A(2)"]);
    }
  },
  {
    level: 3,
    name: "processa por horário e depois ordem numérica de criação",
    run: () => {
      const service = new ParcelLockerService();
      service.addLocker("A", 50);
      service.addLocker("B", 50);
      for (let i = 1; i <= 11; i += 1) {
        service.storeParcel(i, `p${i}`, "A", 1);
        equal(service.scheduleMove(20 + i, `p${i}`, "B", 100), `move-${i}`);
      }
      equal(service.processScheduled(100), [
        "move-1", "move-2", "move-3", "move-4", "move-5", "move-6",
        "move-7", "move-8", "move-9", "move-10", "move-11"
      ]);
    }
  },
  {
    level: 3,
    name: "falha agendada por capacidade é definitiva",
    run: () => {
      const service = new ParcelLockerService();
      service.addLocker("A", 10);
      service.addLocker("B", 3);
      service.storeParcel(1, "p", "A", 4);
      equal(service.scheduleMove(2, "p", "B", 5), "move-1");
      equal(service.processScheduled(5), []);
      equal(service.processScheduled(6), []);
      equal(service.collectParcel(7, "p"), true);
      equal(service.cancelScheduledMove(8, "move-1"), false);
    }
  },
  {
    level: 4,
    name: "fusão sem capacidade falha atomicamente",
    run: () => {
      const service = new ParcelLockerService();
      service.addLocker("A", 5);
      service.addLocker("B", 5);
      service.storeParcel(1, "a", "A", 4);
      service.storeParcel(2, "b", "B", 3);
      equal(service.mergeLockers(3, "A", "B"), false);
      equal(service.getLoad(4, "A"), 4);
      equal(service.getLoad(5, "B"), 3);
      equal(service.topLockers(6, 2), ["A(4)", "B(3)"]);
      equal(service.getParcelHistory(7, "a"), ["stored:A@1"]);
    }
  },
  {
    level: 4,
    name: "fusão redireciona destino de agendamento pendente",
    run: () => {
      const service = new ParcelLockerService();
      service.addLocker("A", 10);
      service.addLocker("B", 10);
      service.addLocker("C", 10);
      service.storeParcel(1, "p", "C", 2);
      equal(service.scheduleMove(2, "p", "A", 10), "move-1");
      equal(service.mergeLockers(3, "A", "B"), true);
      equal(service.processScheduled(10), ["move-1"]);
      equal(service.getLoad(11, "B"), 2);
      equal(service.getParcelHistory(12, "p"), ["stored:C@1", "moved:C->B@10"]);
    }
  }
];
